### PR TITLE
[CALCITE] Add support for NUMBER data type

### PR DESCRIPTION
**Pull request by michaelvdow**
Syntax added: `NUMBER(\((n|*)(,m)?\))?`

### CODE REVIEW

File: core/src/main/java/org/apache/calcite/sql/type/SqlTypeName.java [hunk 1]
```diff
@@ -61,6 +61,7 @@
   FLOAT(PrecScale.NO_NO, false, Types.FLOAT, SqlTypeFamily.NUMERIC),
   REAL(PrecScale.NO_NO, false, Types.REAL, SqlTypeFamily.NUMERIC),
   DOUBLE(PrecScale.NO_NO, false, Types.DOUBLE, SqlTypeFamily.NUMERIC),
+  NUMBER(PrecScale.NO_NO, false, Types.NUMERIC, SqlTypeFamily.NUMERIC),
```
Comment (AndrewPochapsky) [inline core/src/main/java/org/apache/calcite/sql/type/SqlTypeName.java:64]: Is it necessary to add this here? I don't see it being referenced anywhere else in the change.

Comment (michaelvdow) [reply core/src/main/java/org/apache/calcite/sql/type/SqlTypeName.java:64]: Ah good catch, I was just basing it off the VARBYTE one and forgot I didn't use it here. Will remove

File: core/src/main/java/org/apache/calcite/sql/SqlNumberTypeNameSpec.java [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+    return numericPrecision >= 1 && numericPrecision <= 38;
+  }
+
+  private boolean isValidScale(boolean isPrecisionStar, SqlLiteral precision,
+      SqlLiteral scale) {
+    if (scale == null) {
+      return true;
+    }
+    int numericScale = scale.getValueAs(Integer.class);
+    if (isPrecisionStar) {
```
Comment (dasch-1) [inline core/src/main/java/org/apache/calcite/sql/SqlNumberTypeNameSpec.java:77]: nit: Suggest simplifying this logic slightly to:

int maxNumericPrecision = isPrecisionStar ? 38 : numericPrecision;
return numericScale >= 0 && numericScale <= maxNumericPrecision

File: core/src/main/java/org/apache/calcite/sql/SqlNumberTypeNameSpec.java [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+    }
+  }
+
+  private boolean isValidPrecision(boolean isPrecisionStar,
+      SqlLiteral precision) {
+    if (isPrecisionStar || precision == null) {
+      return true;
+    }
+    int numericPrecision = precision.getValueAs(Integer.class);
+    return numericPrecision >= 1 && numericPrecision <= 38;
```
Comment (dasch-1) [inline core/src/main/java/org/apache/calcite/sql/SqlNumberTypeNameSpec.java:68]: Let's define the lower bound and upper bounds in private static final class fields and then reuse here and elsewhere.

File: core/src/main/java/org/apache/calcite/sql/SqlNumberTypeNameSpec.java [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+      precision.unparse(writer, 0, 0);
+    }
+    if (scale != null) {
+      writer.sep(",", true);
+      scale.unparse(writer, 0, 0);
+    }
+    writer.endList(frame);
+  }
+
+  @Override public boolean equalsDeep(SqlTypeNameSpec spec, Litmus litmus) {
```
Comment (dasch-1) [inline core/src/main/java/org/apache/calcite/sql/SqlNumberTypeNameSpec.java:108]: What happens if we just replace all of this logic with return false? Do any tests fail?

If not, I'd suggest just doing that. It's usually better to have no logic at all than logic that's untested.

Comment (michaelvdow) [reply core/src/main/java/org/apache/calcite/sql/SqlNumberTypeNameSpec.java:107]: Yeah it still passes all the tests, I'll change it to just false then

File: core/src/main/java/org/apache/calcite/sql/SqlNumberTypeNameSpec.java [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+import org.apache.calcite.util.Litmus;
+
+import static org.apache.calcite.util.Static.RESOURCE;
+
+/**
+ * A SQL type name specification for the NUMBER data type.
+ */
+public class SqlNumberTypeNameSpec extends SqlTypeNameSpec {
+
+  private final static int MIN_PRECISION = 1;
```
Comment (dasch-1) [inline core/src/main/java/org/apache/calcite/sql/SqlNumberTypeNameSpec.java:31]: "private static final" (here and the next two lines)

Comment (michaelvdow) [reply core/src/main/java/org/apache/calcite/sql/SqlNumberTypeNameSpec.java]: Whoops, got it switched. It's updated

**Comment by dasch-1** [comment]
The macOS test failure says something about timing out in a JDBC unit test. I think we can safely assume that's just a test infrastructure flake.

**Pull request merged**